src/main/java/org/springframework/data/redis/support/collections/DefaultRedisList.java

```diff
@@ -592,4 +582,122 @@ private void cap() {
 			listOps.trim(0, maxSize - 1);
 		}
 	}
+
+	private class DefaultRedisListIterator extends RedisIterator<E> {
+
+		public DefaultRedisListIterator(Iterator<E> delegate) {
+			super(delegate);
+		}
+
+		@Override
+		protected void removeFromRedisStorage(E item) {
+			DefaultRedisList.this.remove(item);
+		}
+	}
+
+	private class Itr implements Iterator<E> {
+
+		/**
+		 * Index of element to be returned by subsequent call to next.
+		 */
+		int cursor = 0;
+
+		@Nullable
+		E lastReturnedElement;
+
+		@Override
+		public boolean hasNext() {
+			return this.cursor < size();
+		}
+
+		@Override
+		public E next() {
+
+			try {
+				int index = this.cursor;
+				this.lastReturnedElement = get(index);
+				this.cursor = index + 1;
+				return this.lastReturnedElement;
+			} catch (IndexOutOfBoundsException cause) {
+				throw new NoSuchElementException(cause);
+			}
+		}
+
+		@Override
+		@SuppressWarnings("all")
+		public void remove() {
+
+			Assert.state(this.lastReturnedElement != null,
+				"Next must be called before remove");
+
+			if (!DefaultRedisList.this.remove(this.lastReturnedElement)) {
+				throw new ConcurrentModificationException();
+			}
+
+			this.lastReturnedElement = null;
+			this.cursor--;
+		}
+	}
+
+	private class ListItr extends Itr implements ListIterator<E> {
+
+		ListItr(int index) {
+			this.cursor = index;
+		}
+
+		@Override
+		public boolean hasPrevious() {
+			return this.cursor > 0;
+		}
+
+		@Override
+		public int nextIndex() {
+			return this.cursor;
+		}
+
+		@Override
+		public int previousIndex() {
+			return this.cursor - 1;
+		}
+
+		@Override
+		public void add(E element) {
+
+			try {
+				int index = this.cursor;
+				DefaultRedisList.this.add(index, element);
+				this.lastReturnedElement = null;
+				this.cursor = index + 1;
+			} catch (IndexOutOfBoundsException cause) {
+				throw new ConcurrentModificationException();
+			}
+		}
+
+		@Override
+		public E previous() {
+
+			try {
+				int index = this.cursor - 1;
+				this.lastReturnedElement = get(index);
+				this.cursor = index;
+				return this.lastReturnedElement;
+			} catch (IndexOutOfBoundsException cause) {
+				throw new NoSuchElementException();
+			}
+		}
+
+		@Override
+
+		public void set(E element) {
+
+			Assert.state(this.lastReturnedElement != null,
+				"Next or previous must be called before set");
+
+			try {
+				DefaultRedisList.this.set(indexOf(this.lastReturnedElement), element);
```

Comment:
Let's resort to using the cursor index. Using the index of the last returned element can hit an equal object in the list (e.g. if the list contains `[a a a b c c c]` and the last returned element is `c`).